stm/stm.ml

```diff
@@ -1319,27 +1285,73 @@ end = struct (* {{{ *)
 
   let get_prev_proof ~doc id =
     try
-      let np = get_proof ~doc id in
-      match np with
-      | None -> None
-      | Some cp ->
-        let did = ref id in
-        let rv = ref np in
-        let done_ = ref false in
-        while not !done_ do
-          did := fold_until back_tactic 1 !did;
-          rv := get_proof ~doc !did;
-          done_ := match !rv with
-            | Some rv -> not (Goal.Set.equal (Proof.all_goals rv) (Proof.all_goals cp))
-            | None -> true
-        done;
-        !rv
+      (* "Show" doesn't have a proof when it calls here; handle it *)
+      let did = ref (if get_proof ~doc id <> None then id
+        else
+          fold_until (fun n (id,_,_,_,_) ->
+                      if Int.equal n 0 then `Stop id else `Cont (n-1)) 1 id) in
+      let np = get_proof ~doc !did in
+      let rv = ref None in
+      let done_ = ref false in
+      while not !done_ do
+        did := fold_until back_tactic 1 !did;
+        rv := get_proof ~doc !did;
+        done_ := match np, !rv with
+          | Some rvp, Some npp -> not (Goal.Set.equal (Proof.all_goals rvp) (Proof.all_goals npp))
+          | _, _ -> !did = Stateid.initial
+      done;
+      !rv
     with Not_found | PG_compat.NoCurrentProof -> None
 
 end (* }}} *)
 
 let get_prev_proof = Backtrack.get_prev_proof
 
+(* Wrapper for Vernacentries.interp to set the feedback id *)
+(* It is currently called 19 times, this number should be certainly
+   reduced... *)
+let stm_vernac_interp ?proof ?route id st { verbose; expr } : Vernacstate.t =
+  (* The Stm will gain the capability to interpret commmads affecting
+     the whole document state, such as backtrack, etc... so we start
+     to design the stm command interpreter now *)
+  set_id_for_feedback ?route dummy_doc id;
+  Aux_file.record_in_aux_set_at ?loc:expr.CAst.loc ();
+  (* We need to check if a command should be filtered from
+   * vernac_entries, as it cannot handle it. This should go away in
+   * future refactorings.
+  *)
+  let is_filtered_command = function
+    | VernacResetName _ | VernacResetInitial | VernacBack _
+    | VernacBackTo _ | VernacRestart | VernacUndo _ | VernacUndoTo _
+    | VernacAbortAll | VernacAbort _ -> true
+    | _ -> false
+  in
+  let aux_interp st expr =
+    (* XXX unsupported attributes *)
+    let cmd = Vernacprop.under_control expr in
+    if is_filtered_command cmd then
+      (stm_pperr_endline Pp.(fun () -> str "ignoring " ++ Ppvernac.pr_vernac expr); st)
+    else
+      match cmd with
+      | VernacShow ShowScript -> ShowScript.show_script (); st (* XX we are ignoring control here *)
+      | _ ->
+        stm_pperr_endline Pp.(fun () -> str "interpreting " ++ Ppvernac.pr_vernac expr);
+        let oproof = match CAst.with_val (fun x -> x) expr with
```

Comment:
I think you should match on `cmd` here

Reply:
Good point, will do.

> I believe the asymmetry in the type of xproof deserves a comment.

Not sure what you mean by this.

Reply:
And where would you put that comment?

Reply:
Next to the type declaration, where I saw that the two components of the pair of "states" being compared are not of the same type. Before the argument was `~proof:Proof.t`, now it is `~xproof:...`